datafusion/core/src/row/layout.rs

```diff
@@ -50,13 +118,23 @@ fn type_width(dt: &DataType) -> usize {
     }
 }
 
+fn word_aligned_offsets(null_width: usize, schema: &Arc<Schema>) -> (Vec<usize>, usize) {
+    let mut offsets = vec![];
+    let mut offset = null_width;
+    for _ in schema.fields() {
+        offsets.push(offset);
+        offset += 8; // a 8-bytes word for each field
```

Comment:
Are we sure that all field types can fit within 8 bytes?  I am thinking about types like `DataType::Decimal` stored in a `u128` / `16 bytes`)?

Also it would be nice to know variable length data works here

Reply:
Row format doesn't support Decimal for now, so just 8 bytes currently.

Reply:
It just seemed like a landmine for the future. It is fine

Reply:
Let me make it a function here to choose between 8 and 16 based on DataTypes. 

> Also it would be nice to know variable length data works here

I haven't settled down my mind if I'd store grouping keys as well as group state as one row or two separate rows and have a JoinedRow { key: vec<u8>, state: vec<u8> } for the hashtable in aggregate.

But for the group states only, I prefer to not handle variable length states, such as an array, median, or sketch in approxs', since it's not efficient to do in-place updates for these states, we'd better keep with Vec<ScalarValue> states for them.

Reply:
Sorry -- I should have been clearer -- what I think is important is that the code will assert / fail in some obvious way if/when someone decides to add `DataType::Decimal` support in the future. I don't think we actually need to support any additional sizes now, I am just thinking of our future selves (or some other new contributor) trying to debug a partial field overwrite because the fields were too small 🤯